### PR TITLE
[rocprofiler-systems] Replace get_env and set_env from timemory inside rocprofiler-systems

### PR DESCRIPTION
**Pull request by mradosav-amd**
## Motivation

Remove the dependency on `tim::get_env` / `tim::set_env` / `tim::get_env_choice` from `source/lib/` and `source/python/`, replacing them with the project-owned `rocprofsys::common::get_env` / `set_env` / `get_env_choice` from `source/lib/common/environment.hpp`. This keeps timemory as a build dependency only for the components that genuinely need it, and gives the project full control over how environment variables are read.

A secondary goal was to make the environment layer unit-testable without touching the real process environment, which drove the backend-parameterisation refactor described below.

### Replace `tim::get_env` / `tim::set_env` with `rocprofsys::common` equivalents

**`source/lib/common/environment.hpp`**
- Parses integral and floating-point types with `std::from_chars` (locale-independent, non-throwing): negative or out-of-range input for the target type falls back to the default instead of wrapping or truncating, and a very large all-digit boolean value no longer throws
- `env_id` is `const char*`-oriented (primary, allocation-free overload) with a `std::string_view` / `std::string` shim, so the upcoming `constexpr const char*` env_vars constants read without allocation
- Added 1-argument `get_env<Tp>(env_id)` overload returning `Tp{}` as default
- Added `get_env_choice(env_id, default, std::set<Tp> choices)` — validates the env value against an allowed set and falls back to the default on a bad value

**`source/lib/core/config.hpp`**
- Added `#include "common/environment.hpp"` so the entire core library gets `rocprofsys::common::get_env` in scope through a single include

**`source/lib/core/mpi.hpp`**
- Replaced `#include <timemory/environment/declaration.hpp>` with `#include "common/environment.hpp"`

**`source/lib/core/dynamic_library.hpp` / `.cpp`**
- Removed the `bool _store` parameter from the `dynamic_library` constructor — it was only forwarded to `tim::get_env` and has no meaning in the common API

**Remaining files** (`config.cpp`, `perfetto.cpp`, `trace_cache/discovery.cpp`, `library.cpp`, `timeout.cpp`, `causal/data.cpp`, `runtime.cpp`, `dynamic_library.cpp`, `libpyrocprofsys.cpp`)
- Removed `#include <timemory/environment.hpp>` / `<timemory/environment/types.hpp>` where only present for `get_env`
- Stripped the `_store` boolean (third argument of `tim::get_env`) from all call sites
- Replaced `get_env_choice<T>(..., choices, false)` with `get_env_choice<T>(..., choices)` using the new common implementation

### Link the instrumentation tool against the common library

**`source/bin/rocprof-sys-instrument/CMakeLists.txt`**
- Replaced the direct `rocprofiler-systems::rocprofiler-systems-spdlog` link with `rocprofiler-systems::rocprofiler-systems-common-library`, which transitively provides the logger/spdlog now used by `environment.hpp` and aligns this tool with the other `source/bin/` targets

The remaining `source/bin/` tools (`rocprof-sys-avail`, `rocprof-sys-causal`) continue to use `tim::get_env` via their own `using ::tim::get_env` declarations and were not changed in this PR. `rocprof-sys-instrument` only changed its link line (above) to pull the logging dependency from the common library.

No behaviour change is intended — `rocprofsys::common::get_env` is a plain `getenv` wrapper. The only observable difference from `tim::get_env` is that lookups are no longer recorded in timemory's `env_settings` registry, which is a timemory-internal concern.

### Refactor environment layer to be mockable and unit-testable

**`source/lib/common/environment.hpp`**
- Introduced a `posix_env` backend struct (static `getenv` / `setenv` forwarding to `::`) representing the real POSIX environment
- Refactored all `get_env_impl` / `set_env` / `update_env` / `remove_env` logic into an `environment<EnvType = posix_env>` template struct, parameterised over the backend
- Free functions `get_env`, `set_env`, `update_env`, `remove_env` remain as thin wrappers over `environment<posix_env>` — call sites are unchanged
- A test-only fake backend (e.g. `struct fake_env { static std::map<...> store; ... }`) can be injected without spawning a child process or polluting the real process environment

**`source/lib/common/tests/test_environment.cpp`**
- Rewrote test suites to use the injected fake backend, eliminating all `setenv`/`unsetenv` calls that could affect other tests running in the same process
- Added coverage for: `get_env` (string, int, bool, `size_t`, `double`, 1-arg form), `set_env` (override semantics), `get_env_choice` (valid/invalid values)

### Fix Python extension undefined symbol

**`source/python/libpyrocprofsys.cpp`**
- `libpyrocprofsys.so` had an undefined symbol for `tim::get_env<bool>` after removing `tim::get_env` usages from `librocprof-sys.so`: `timemory/mpl/policy.hpp` pulls in `type_traits.hpp` which invokes `get_env<bool>` via `TIMEMORY_REPORT_ENV_QUERY`, with only `environment/types.hpp` (declaration) visible — the inline definition had previously been dragged in transitively through the main library's symbol table. Fix: include `timemory/environment/definition.hpp` before `mpl/policy.hpp` so the specialisation definition is compiled directly into the extension.

**`source/lib/common/defines.h.in`**
- Fixed macro alignment whitespace flagged by `clang-format-18`

## Test Plan

- Full debug build (`cmake --build build/debug -j $(nproc)`) with all 536 targets
- Unit tests in `source/lib/common/tests/test_environment.cpp` extended and rewritten to use injected fake backend

## Test Result

- Build: 536/536 targets, zero errors
- Unit tests: `unit-tests` CTest target — 100% pass (0 failures)

## Submission Checklist

- [ ] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

### CODE REVIEW

**Review by marantic-amd — APPROVED** [review]

File: projects/rocprofiler-systems/source/lib/common/environment.hpp [hunk 1]
```diff
@@ -67,144 +74,285 @@ namespace rocprofsys
...
+    }
+
+    template <typename Tp>
+    static auto get_env_choice(std::string_view env_id, Tp value_default,
+                               std::set<Tp> choices)
+    {
+        auto value = get_env(env_id, value_default);
+        if(choices.find(value) == choices.end())
+        {
+            fprintf(
```
Comment (adjordje-amd) [inline projects/rocprofiler-systems/source/lib/common/environment.hpp]: Use the existing logging system instead of fprintf

Comment (mradosav-amd) [reply projects/rocprofiler-systems/source/lib/common/environment.hpp]: Resolved.

File: projects/rocprofiler-systems/source/lib/common/environment.hpp [hunk 2]
```diff
@@ -4,207 +4,305 @@
...
+        if(!raw) return fallback;
+        try
+        {
+            if constexpr(std::is_same_v<Tp, int>)
+            {
+                return std::stoi(raw);
+            }
+            else
+            {
+                return static_cast<Tp>(std::stoll(raw));
```
Comment (habajpai-amd) [inline projects/rocprofiler-systems/source/lib/common/environment.hpp]: This needs separate handling for unsigned integral types. Right now every non-int integral goes through `std::stoll(raw)` and then `static_cast<Tp>`, so `get_env<std::uint64_t>(..., 1)` with -1 returns `UINT64_MAX` instead of falling back to the default. The same applies to `std::uint16_t` and `size_t` call sites such as `ROCPROFSYS_CI_TIMEOUT_COUNT`, `ROCPROFSYS_CAUSAL_SPEEDUP_DIVISIONS`, and `ROCPROFSYS_NUM_THREADS`.
Could we reject leading `-` for unsigned types, parse with `std::stoull` or `std::from_chars`, and range-check before casting? Please also add tests for negative unsigned input and out-of-range unsigned input.

File: projects/rocprofiler-systems/source/bin/rocprof-sys-instrument/CMakeLists.txt [hunk 3]
```diff
@@ -37,6 +37,7 @@ target_link_libraries(
         rocprofiler-systems::rocprofiler-systems-compile-options
         rocprofiler-systems::rocprofiler-systems-compile-definitions
         rocprofiler-systems::rocprofiler-systems-sanitizer
+        rocprofiler-systems::rocprofiler-systems-spdlog
```
Comment (sputhala-amd) [inline projects/rocprofiler-systems/source/bin/rocprof-sys-instrument/CMakeLists.txt]: This could also be replaced with `rocprofiler-systems::rocprofiler-systems-common-library` which aligns with other places too.

Comment (sputhala-amd) [reply projects/rocprofiler-systems/source/bin/rocprof-sys-instrument/CMakeLists.txt]: The link to `spdlog` will work too, just a minor suggestion

File: projects/rocprofiler-systems/source/lib/common/environment.hpp [hunk 4]
```diff
@@ -4,207 +4,313 @@
...
+
+#include <timemory/utility/filepath.hpp>
+
 #include <algorithm>
+#include <array>
 #include <cctype>
+#include <charconv>
 #include <cstdint>
 #include <cstdio>
 #include <cstdlib>
```
Comment (dgaliffiAMD) [inline projects/rocprofiler-systems/source/lib/common/environment.hpp]:
```suggestion
```
<stdlib.h> will include the `::setenv` and `::getenv`.

File: projects/rocprofiler-systems/source/lib/common/environment.hpp [hunk 5]
```diff
@@ -4,207 +4,313 @@
...
+        const std::string_view env_sv{ raw };
+        if(env_sv.empty())
+        {
+            throw std::runtime_error(
+                std::string{ "No boolean value provided for " }.append(env_id));
+        }
+
+        if(env_sv.find_first_not_of("0123456789") == std::string_view::npos)
+        {
+            return static_cast<bool>(std::stoi(raw));
```
Comment (dgaliffiAMD) [inline projects/rocprofiler-systems/source/lib/common/environment.hpp]:
```suggestion
            int value{};
            auto [ptr, ec] = std::from_chars(env_sv.data(), env_sv.data() + env_sv.size(), value);
            if(ec == std::errc{}) return value != 0;
            return true;
```
A very large integer can throw `std::out_of_range` and this is not caught. Later on, in this file, you use `std::from_chars`. Can you use it here too?

File: projects/rocprofiler-systems/source/lib/common/environment.hpp [hunk 6]
```diff
@@ -4,207 +4,313 @@
...
+    template <typename Tp>
+    static auto get_env_choice(std::string_view env_id, Tp value_default,
+                               std::set<Tp> choices)
+    {
+        auto value = get_env(env_id, value_default);
+        if(choices.find(value) == choices.end())
+        {
+            LOG_WARNING("[get_env] Environment variable \"{}\" has invalid value \"{}\". "
+                        "Reverting to default.",
+                        env_id, get_env<std::string>(env_id));
```
Comment (dgaliffiAMD) [inline projects/rocprofiler-systems/source/lib/common/environment.hpp]: nit-pick: on the error, can you output the cached `value` or `fetch_raw_env` to avoid calling the full parsing again?

File: projects/rocprofiler-systems/source/lib/common/environment.hpp [hunk 7]
```diff
@@ -4,207 +4,313 @@
...
 {
-    if(env_id.empty()) return _default;
-    char* env_var = ::std::getenv(env_id.data());
-    if(env_var)
+private:
+    // env_id is always constructed from a null-terminated string at every call site
+    static const char* fetch_raw_env(std::string_view env_id)
+    {
+        if(env_id.empty()) return nullptr;
+        return EnvType::getenv(
```
Comment (dgaliffiAMD) [inline projects/rocprofiler-systems/source/lib/common/environment.hpp]: I know that you describe in the comments why the linting error is being ignored, but why can't you use a `const char*` for `env_id` instead?

```
    static const char* fetch_raw_env(const char* env_id)
    {
        if(env_id == nullptr || env_id[0] == '\0') return nullptr;
        return EnvType::getenv(env_id);
    }
```

File: projects/rocprofiler-systems/source/bin/rocprof-sys-instrument/CMakeLists.txt (hunk 3 above)
Comment (dgaliffiAMD) [reply projects/rocprofiler-systems/source/bin/rocprof-sys-instrument/CMakeLists.txt]: Maybe a small note in the PR description about updating the logging here too.

**Comment by sputhala-amd** [comment]
LGTM. There is a workflow failure due to another project (compute component). I have re-run the workflow.
Leaving for others to have a look and approve.

**Review by dgaliffiAMD — APPROVED** [review]
Thank you.
